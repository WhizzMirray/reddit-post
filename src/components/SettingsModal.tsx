import React from "react";
import { Checkbox, FormGroup, Modal } from "react-bootstrap";
import { connect } from "react-redux";

import { Dispatch } from "redux";
import {
  changeDelay,
  toggleNotification,
  toggleSound
} from "../actions/settingsActions";
import { closeSettingsModal } from "../actions/utilActions";
import { IState } from "../types/interfaces";

interface IProps {
  showModal: any;
  settings: any;
}

interface IOwnState {
  permission: string;
}

interface IDispatchProps {
  handleChangeDelay: (e: Event & any) => void;
  handleCloseSettingsModal: () => void;
  handleToggleNotification: () => Promise<any>;
  handleToggleSound: () => void;
}

type Props = IProps & IDispatchProps;

class SettingsModal extends React.Component<Props, IOwnState> {
  constructor(props: Props) {
    super(props);

    this.state = {
      permission: (window as any).Notification
        ? (window as any).Notification.permission
        : "default"
    };
    this.handleToggleNotification = this.handleToggleNotification.bind(this);
  }
  
  public render() {
    const { delay, sound, notification } = this.props.settings;
    const {
      handleChangeDelay,
      handleCloseSettingsModal,
      handleToggleSound,
      showModal
    } = this.props;
    return (
      <div>
        <Modal show={showModal} onHide={handleCloseSettingsModal}>
          <Modal.Header closeButton={true}>
            <Modal.Title>Settings</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <FormGroup>
              <Checkbox checked={sound} onChange={handleToggleSound}>
                Toggle Sound
              </Checkbox>{" "}
              {(window as any).Notification ? (
                <Checkbox
                  checked={notification}
                  onChange={this.handleToggleNotification}
                  disabled={this.state.permission === "denied"}
                >
                  Toggle Notifications
                </Checkbox>
              ) : (
                ""
              )}{" "}
              {"Delay "}
              <input
                type="number"
                value={delay}
                onChange={handleChangeDelay}
                min={5}
              />
              {" s"}
            </FormGroup>
          </Modal.Body>
        </Modal>
      </div>
    );
  }
  // handleToggleNotification wrapper to use this.setState
  public handleToggleNotification() {
    this.props
      .handleToggleNotification()
      .then(() => this.setState({ permission: "granted" }));
  }
}
const mapStateToProps = (state: IState): IProps => ({
  settings: state.settings,
  showModal: state.utils.settingsModalOpen
});

const mapDispatchToProps = (dispatch: Dispatch<any>): IDispatchProps => ({
  handleChangeDelay: e => {
    e.preventDefault();
    const n = Number.parseInt(e.target.value, 0);
    dispatch(changeDelay(n));
  },
  handleCloseSettingsModal: () => {
    dispatch(closeSettingsModal());
  },
  handleToggleNotification: () => {
    return Notification.requestPermission().then(result => {
      if (result === "granted") {
        dispatch(toggleNotification());
      }
    });
  },
  handleToggleSound: () => {
    dispatch(toggleSound());
  }
});

export default connect<IProps, IDispatchProps>(
  mapStateToProps,
  mapDispatchToProps
)(SettingsModal);
