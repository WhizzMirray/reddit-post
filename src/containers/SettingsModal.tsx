import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import {
  changeDelay,
  toggleNotification,
  toggleSound
} from "../actions/settingsActions";
import { closeSettingsModal } from "../actions/utilActions";

import { ISettings, IState } from "../types/interfaces";

import SettingsModalComponent from "../components/SettingsModal";

interface IStateProps {
  showModal: boolean;
  settings: ISettings;
}

interface IOwnState {
  permission: string;
}

interface IDispatchProps {
  handleChangeDelay: (e: Event & any) => void;
  handleCloseSettingsModal: () => void;
  handleToggleNotification: () => Promise<any>;
  handleToggleSound: () => void;
}

type Props = IStateProps & IDispatchProps;

class SettingsModal extends React.Component<Props, IOwnState> {
  constructor(props: Props) {
    super(props);

    this.state = {
      permission: (window as any).Notification
        ? (window as any).Notification.permission
        : "default"
    };
    this.handleToggleNotification = this.handleToggleNotification.bind(this);
  }

  // handleToggleNotification wrapper to use this.setState
  public handleToggleNotification() {
    this.props
      .handleToggleNotification()
      .then(() => this.setState({ permission: "granted" }));
  }
  public render() {
    return (
      <SettingsModalComponent
        {...this.state}
        {...this.props}
        handleToggleNotification={this.handleToggleNotification}
      />
    );
  }
}

const mapStateToProps = (state: IState): IStateProps => ({
  settings: state.settings,
  showModal: state.utils.settingsModalOpen
});

const mapDispatchToProps = (dispatch: Dispatch<any>): IDispatchProps => ({
  handleChangeDelay: e => {
    e.preventDefault();
    const n = Number.parseInt(e.target.value, 0);
    dispatch(changeDelay(n));
  },
  handleCloseSettingsModal: () => {
    dispatch(closeSettingsModal());
  },
  handleToggleNotification: () => {
    return Notification.requestPermission().then(result => {
      if (result === "granted") {
        dispatch(toggleNotification());
      }
    });
  },
  handleToggleSound: () => {
    dispatch(toggleSound());
  }
});

export default connect<IStateProps, IDispatchProps>(
  mapStateToProps,
  mapDispatchToProps
)(SettingsModal);
