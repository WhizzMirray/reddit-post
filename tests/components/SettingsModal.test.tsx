import { shallow } from "enzyme";
import * as React from "react";

import SettingsModal, { IProps } from "../../src/components/SettingsModal";

const props : IProps = {
  handleChangeDelay: jest.fn(),
  handleCloseSettingsModal: jest.fn(),
  handleToggleNotification: jest.fn(),
  handleToggleSound: jest.fn(),
  permission: "allowed",
  settings: {
    delay: 5,
    notification: true,
    sound: true
  },
  showModal: true
};
describe("<SettingsModal/>", () => {
  it("should render correctly", () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
