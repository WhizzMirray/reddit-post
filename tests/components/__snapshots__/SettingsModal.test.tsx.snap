// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsModal/> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SettingsModal
    handleChangeDelay={[MockFunction]}
    handleCloseSettingsModal={[MockFunction]}
    handleToggleNotification={[MockFunction]}
    handleToggleSound={[MockFunction]}
    permission="allowed"
    settings={
      Object {
        "delay": 5,
        "notification": true,
        "sound": true,
      }
    }
    showModal={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Modal
        animation={true}
        autoFocus={true}
        backdrop={true}
        bsClass="modal"
        dialogComponentClass={[Function]}
        enforceFocus={true}
        keyboard={true}
        manager={
          ModalManager {
            "add": [Function],
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "isTopModal": [Function],
            "modals": Array [],
            "remove": [Function],
          }
        }
        onHide={[MockFunction]}
        renderBackdrop={[Function]}
        restoreFocus={true}
        show={true}
      >
        <ModalHeader
          bsClass="modal-header"
          closeButton={true}
          closeLabel="Close"
        >
          <ModalTitle
            bsClass="modal-title"
            componentClass="h4"
          >
            Settings
          </ModalTitle>
        </ModalHeader>
        <ModalBody
          bsClass="modal-body"
          componentClass="div"
        >
          <FormGroup
            bsClass="form-group"
          >
            <Checkbox
              bsClass="checkbox"
              checked={true}
              disabled={false}
              inline={false}
              onChange={[MockFunction]}
              title=""
            >
              Toggle Sound
            </Checkbox>
             
            
             Delay  
            <input
              min={5}
              onChange={[MockFunction]}
              type="number"
              value={5}
            />
             s
          </FormGroup>
        </ModalBody>
      </Modal>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "animation": true,
        "autoFocus": true,
        "backdrop": true,
        "bsClass": "modal",
        "children": Array [
          <ModalHeader
            bsClass="modal-header"
            closeButton={true}
            closeLabel="Close"
          >
            <ModalTitle
              bsClass="modal-title"
              componentClass="h4"
            >
              Settings
            </ModalTitle>
          </ModalHeader>,
          <ModalBody
            bsClass="modal-body"
            componentClass="div"
          >
            <FormGroup
              bsClass="form-group"
            >
              <Checkbox
                bsClass="checkbox"
                checked={true}
                disabled={false}
                inline={false}
                onChange={[MockFunction]}
                title=""
              >
                Toggle Sound
              </Checkbox>
               
              
               Delay  
              <input
                min={5}
                onChange={[MockFunction]}
                type="number"
                value={5}
              />
               s
            </FormGroup>
          </ModalBody>,
        ],
        "dialogComponentClass": [Function],
        "enforceFocus": true,
        "keyboard": true,
        "manager": ModalManager {
          "add": [Function],
          "containers": Array [],
          "data": Array [],
          "handleContainerOverflow": true,
          "hideSiblingNodes": true,
          "isTopModal": [Function],
          "modals": Array [],
          "remove": [Function],
        },
        "onHide": [MockFunction],
        "renderBackdrop": [Function],
        "restoreFocus": true,
        "show": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "modal-header",
            "children": <ModalTitle
              bsClass="modal-title"
              componentClass="h4"
            >
              Settings
            </ModalTitle>,
            "closeButton": true,
            "closeLabel": "Close",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "modal-title",
              "children": "Settings",
              "componentClass": "h4",
            },
            "ref": null,
            "rendered": "Settings",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "modal-body",
            "children": <FormGroup
              bsClass="form-group"
            >
              <Checkbox
                bsClass="checkbox"
                checked={true}
                disabled={false}
                inline={false}
                onChange={[MockFunction]}
                title=""
              >
                Toggle Sound
              </Checkbox>
               
              
               Delay  
              <input
                min={5}
                onChange={[MockFunction]}
                type="number"
                value={5}
              />
               s
            </FormGroup>,
            "componentClass": "div",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <Checkbox
                  bsClass="checkbox"
                  checked={true}
                  disabled={false}
                  inline={false}
                  onChange={[MockFunction]}
                  title=""
                >
                  Toggle Sound
                </Checkbox>,
                " ",
                "",
                " Delay  ",
                <input
                  min={5}
                  onChange={[MockFunction]}
                  type="number"
                  value={5}
                />,
                " s",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "checkbox",
                  "checked": true,
                  "children": "Toggle Sound",
                  "disabled": false,
                  "inline": false,
                  "onChange": [MockFunction],
                  "title": "",
                },
                "ref": null,
                "rendered": "Toggle Sound",
                "type": [Function],
              },
              " ",
              "",
              " Delay  ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "min": 5,
                  "onChange": [MockFunction],
                  "type": "number",
                  "value": 5,
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              " s",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          bsClass="modal"
          dialogComponentClass={[Function]}
          enforceFocus={true}
          keyboard={true}
          manager={
            ModalManager {
              "add": [Function],
              "containers": Array [],
              "data": Array [],
              "handleContainerOverflow": true,
              "hideSiblingNodes": true,
              "isTopModal": [Function],
              "modals": Array [],
              "remove": [Function],
            }
          }
          onHide={[MockFunction]}
          renderBackdrop={[Function]}
          restoreFocus={true}
          show={true}
        >
          <ModalHeader
            bsClass="modal-header"
            closeButton={true}
            closeLabel="Close"
          >
            <ModalTitle
              bsClass="modal-title"
              componentClass="h4"
            >
              Settings
            </ModalTitle>
          </ModalHeader>
          <ModalBody
            bsClass="modal-body"
            componentClass="div"
          >
            <FormGroup
              bsClass="form-group"
            >
              <Checkbox
                bsClass="checkbox"
                checked={true}
                disabled={false}
                inline={false}
                onChange={[MockFunction]}
                title=""
              >
                Toggle Sound
              </Checkbox>
               
              
               Delay  
              <input
                min={5}
                onChange={[MockFunction]}
                type="number"
                value={5}
              />
               s
            </FormGroup>
          </ModalBody>
        </Modal>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animation": true,
          "autoFocus": true,
          "backdrop": true,
          "bsClass": "modal",
          "children": Array [
            <ModalHeader
              bsClass="modal-header"
              closeButton={true}
              closeLabel="Close"
            >
              <ModalTitle
                bsClass="modal-title"
                componentClass="h4"
              >
                Settings
              </ModalTitle>
            </ModalHeader>,
            <ModalBody
              bsClass="modal-body"
              componentClass="div"
            >
              <FormGroup
                bsClass="form-group"
              >
                <Checkbox
                  bsClass="checkbox"
                  checked={true}
                  disabled={false}
                  inline={false}
                  onChange={[MockFunction]}
                  title=""
                >
                  Toggle Sound
                </Checkbox>
                 
                
                 Delay  
                <input
                  min={5}
                  onChange={[MockFunction]}
                  type="number"
                  value={5}
                />
                 s
              </FormGroup>
            </ModalBody>,
          ],
          "dialogComponentClass": [Function],
          "enforceFocus": true,
          "keyboard": true,
          "manager": ModalManager {
            "add": [Function],
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "isTopModal": [Function],
            "modals": Array [],
            "remove": [Function],
          },
          "onHide": [MockFunction],
          "renderBackdrop": [Function],
          "restoreFocus": true,
          "show": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "modal-header",
              "children": <ModalTitle
                bsClass="modal-title"
                componentClass="h4"
              >
                Settings
              </ModalTitle>,
              "closeButton": true,
              "closeLabel": "Close",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "modal-title",
                "children": "Settings",
                "componentClass": "h4",
              },
              "ref": null,
              "rendered": "Settings",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "modal-body",
              "children": <FormGroup
                bsClass="form-group"
              >
                <Checkbox
                  bsClass="checkbox"
                  checked={true}
                  disabled={false}
                  inline={false}
                  onChange={[MockFunction]}
                  title=""
                >
                  Toggle Sound
                </Checkbox>
                 
                
                 Delay  
                <input
                  min={5}
                  onChange={[MockFunction]}
                  type="number"
                  value={5}
                />
                 s
              </FormGroup>,
              "componentClass": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-group",
                "children": Array [
                  <Checkbox
                    bsClass="checkbox"
                    checked={true}
                    disabled={false}
                    inline={false}
                    onChange={[MockFunction]}
                    title=""
                  >
                    Toggle Sound
                  </Checkbox>,
                  " ",
                  "",
                  " Delay  ",
                  <input
                    min={5}
                    onChange={[MockFunction]}
                    type="number"
                    value={5}
                  />,
                  " s",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "bsClass": "checkbox",
                    "checked": true,
                    "children": "Toggle Sound",
                    "disabled": false,
                    "inline": false,
                    "onChange": [MockFunction],
                    "title": "",
                  },
                  "ref": null,
                  "rendered": "Toggle Sound",
                  "type": [Function],
                },
                " ",
                "",
                " Delay  ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "min": 5,
                    "onChange": [MockFunction],
                    "type": "number",
                    "value": 5,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                " s",
              ],
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
